lib/accessmonitoring/monitor.go

```
@@ -0,0 +1,242 @@
+	cfg Config
+
+	// ruleHandlers contains the list of access monitoring rule event handlers.
+	ruleHandlers []EventHandler
+	// requestHandlers contains the list access request event handlers.
+	requestHandlers []EventHandler
+}
+
+// NewAccessMonitor returns a new access monitor.
+func NewAccessMonitor(cfg Config) (*AccessMonitor, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &AccessMonitor{
+		cfg: cfg,
+	}, nil
+}
+
+// AddAccessMonitoringRuleHandler appends the access monitoring rule event handler.
+// These handlers are executed whenever a new access monitoring rule event is
+// observed.
+func (s *AccessMonitor) AddAccessMonitoringRuleHandler(handler EventHandler) {
+	s.ruleHandlers = append(s.ruleHandlers, handler)
+}
+
+// AddAccessRequestHandler appends the access request handler.
+// These handlers are executed whenever a new access request event is observed.
+func (s *AccessMonitor) AddAccessRequestHandler(handler EventHandler) {
+	s.requestHandlers = append(s.requestHandlers, handler)
+}
+
+// Run runs the access monitor. Restarts on failure.
+func (s *AccessMonitor) Run(ctx context.Context) error {
+	const retryJitter = 10 * time.Second
+	const retryInterval = 30 * time.Second
+	const lockTTL = time.Minute
+
+	waitWithJitter := retryutils.SeventhJitter(retryJitter)
+	for {
+		err := backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+			LockConfiguration: backend.LockConfiguration{
+				Backend:            s.cfg.Backend,
+				LockNameComponents: []string{componentName},
+				TTL:                lockTTL,
+				RetryInterval:      retryInterval,
+			},
+		}, s.run)
+		if err != nil {
+			s.cfg.Logger.ErrorContext(
+				ctx,
+				"Encountered a fatal error, it will restart after backoff.",
+				"component", componentName,
+				"error", err,
+				"restart_after", waitWithJitter,
+			)
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(waitWithJitter):
+		}
+	}
+}
```

We had multiple issues with the watcher being stuck to a broken connection and not exiting in kube (I strongly suspect this was caused by a bad grpc connection error detection from our side, and broken kubernetes network sandbox on the user's side).

Because of this we had to introduce a "fail fast" mode when running containerized: https://github.com/gravitational/teleport/pull/30039

I wonder if we should put the env var back in place or if grpc client and watcher improvements will avoid this issue to happen again.

---

Thanks for pointing this out. I'd rather not risk introducing a regression. I'll go ahead and include the "fail fast" mode and include a todo note to investigate if it is still required.